### PR TITLE
Make use of dynamic test registration

### PR DESCRIPTION
**Pull request by fnordahl**
Please review/merge commit by commit.

### CODE REVIEW

File: tests/test_helper/bats/upgrade.bats [hunk 1]
```diff
@@ -1,51 +1,10 @@
...
-    echo "# $output" >&3
-    echo "#" >&3
-    assert_success
-
-    run bats -F junit $tls_test_filename
-
-    echo "# $output" >&3
-    echo "#" >&3
-    assert_success
-}
```
Comment (mkalcok) [inline tests/test_helper/bats/upgrade.bats:51]: I'm just wondering here whether it wouldn't be more readable to keep the `Verify that currently released snap can be upgraded` test and just replace the `run bats -F junit ...` hack with `load test_helper/bats/...`. Are there any benefits of removing this test by moving its "body" to the `setup_file` function?

One other potential drawback would also be less explicit output when running the tests. Running `upgrade.bats` would look like just like running `cluster.bats` and `tls_cluster.bats` combined.

**Review by mkalcok — COMMENTED** [review]
Can this be replaced with dynamic loading as well?
https://github.com/fnordahl/microovn/blob/6ec00ac2a9710fc05dcce3d8073b05c69d86b37c/tests/scaleup_cluster.bats#L67-L75

Or would that just queue the tests at the end and we need, in this instance, run these tests "in-line", during every loop execution?

File: tests/test_helper/bats/upgrade.bats (hunk 1 above)
Comment (fnordahl) [reply tests/test_helper/bats/upgrade.bats:51]: Loading definitions inside a test is unfortunately not a thing.

The setup statements are asserted just like everything else and there is a comment at the very top pointing at where the upgrade steps are happening.

Comment (mkalcok) [reply tests/test_helper/bats/upgrade.bats:51]: > Loading definitions inside a test is unfortunately not a thing.

I wasn't aware of this limitation...then all my questions are moot

**Review by mkalcok — APPROVED** [review]
I really like this change @fnordahl. Untangling the `junit` outputs from the "included" tests was not a pleasant experience. LGTM!

**Comment by fnordahl** [comment]
> Can this be replaced with dynamic loading as well? https://github.com/fnordahl/microovn/blob/6ec00ac2a9710fc05dcce3d8073b05c69d86b37c/tests/scaleup_cluster.bats#L67-L75

Possibly, I looked at it briefly and it has a different mode than the other tests, so I'd save it for a future PR.

> Or would that just queue the tests at the end and we need, in this instance, run these tests "in-line", during every loop execution?

What we potentially can do is use the parametrization and pass a expected number of nodes to the tests, they can then conditionally all call into a `scale_to_n_nodes` helper. The iteration with arg 1 would then do nothing, the first test in the iteration with arg 2 scales the cluster and so on. But, let's save for a future PR.

**Comment by mkalcok** [comment]
> What we potentially can do is use the parametrization and pass a expected number of nodes to the tests, they can then conditionally all call into a scale_to_n_nodes helper. The iteration with arg 1 would then do nothing, the first test in the iteration with arg 2 scales the cluster and so on. But, let's save for a future PR.

I was thinking about exactly the same approach right now. But as you said, that can be done in the future PRs.

**Pull request merged**